File: 1d/ftcs_test.py
# ASTE-404 Homework 8 Part 2
# Colin Coane
# 1D Unsteady Heat Equation Solver, using FTCS scheme

import numpy as np
import matplotlib.pyplot as plt
from random import random as rand
import math

"""
Given parameters:

Domain Length: 1
Number of spatial cells: 60
Number of time steps: 15,000
Time Step: 10^-5
Boundary Conditions: T(t,0) = T(t,1) = 1
Initial Conditions: T(0,x) = 0

"""

ni = 61     # number of nodes
nk = 35001  # number of time steps plus initial time t = 0

dx = 1/(ni - 1) # node spacing
dt = 1e-3       # time step

# Set coefficients
L = 2*math.pi
q = 1.0
alpha = 0.3
uc = 1.1
s = 3.5
k0 = 1.5
eps = 0.15
r = 5
nu1 = 0.0075
nu2 = 0.0075
up = 0.5

# Create array for u and lambda
u = np.zeros((nk,ni))   # np.zeros() sets initial condition to zero
lam = np.zeros((nk,ni)) # np.zeros() sets initial condition to zero

# Initialize pulse characteristics before wave
for i in range(ni):
    u[0,i] = (3/2)*((1/np.cosh(8*(i*dx-1)))**2)
    lam[0,i] = 0.5

#for i in range(ni):
    #print(u[0,i])
    #print(lam[0,i])

# loop over k - march forward in time
for k in range(nk-1):
    #print(k)
    # loop over spatial range
    for i in range(1,ni-1):
        # FTCS algorithm
        # Solve u
        u[k+1,i] = u[k,i] + dt*nu1/(dx*dx)*(u[k,i-1] - 2*u[k,i] + u[k,i+1]) 
        u[k+1,i] += -dt*u[k,i]*(u[k,i+1] - u[k,i-1])/(2*dx) 
        u[k+1,i] += dt*k0*q*(1-lam[k,i])*np.exp((u[k,i]-uc)/alpha)
        u[k+1,i] += -dt*eps*u[k,i]*u[k,i]
        # Solve lambda
        lam[k+1,i] = lam[k,i] + dt*nu2/(dx*dx)*(lam[k,i-1] - 2*lam[k,i] + lam[k,i+1]) 
        lam[k+1,i] += dt*k0*(1-lam[k,i])*np.exp((u[k,i]-uc)/alpha) 
        lam[k+1,i] += -dt*s*up*lam[k,i]/(1 + np.exp(r*(u[k,i] - up)))

    # Apply periodic boundary conditions

    # At i = 0
    # solve u
    u[k+1,0] = u[k,0] + dt*nu1/(dx*dx)*(u[k,ni-1] - 2*u[k,0] + u[k,1]) 
    u[k+1,0] += -dt*u[k,0]*(u[k,1] - u[k,ni-1])/(2*dx) 
    u[k+1,0] += dt*k0*q*(1-lam[k,0])*np.exp((u[k,0]-uc)/alpha) 
    u[k+1,0] += -dt*eps*u[k,0]*u[k,0]
    # Solve lambda
    lam[k+1,0] = lam[k,0] + dt*nu2/(dx*dx)*(lam[k,ni-1] - 2*lam[k,0] + lam[k,1]) 
    lam[k+1,0] += dt*k0*(1-lam[k,0])*np.exp((u[k,0]-uc)/alpha) 
    lam[k+1,0] += -dt*s*up*lam[k,0]/(1 + np.exp(r*(u[k,0] - up)))

    # At i = ni-1
    # Solve u
    u[k+1,ni-1] = u[k,ni-1] + dt*nu1/(dx*dx)*(u[k,ni-2] - 2*u[k,ni-1] + u[k,0]) 
    u[k+1,ni-1] += -dt*u[k,ni-1]*(u[k,0] - u[k,ni-2])/(2*dx) 
    u[k+1,ni-1] += dt*k0*q*(1-lam[k,ni-1])*np.exp((u[k,ni-1]-uc)/alpha) 
    u[k+1,ni-1] += -dt*eps*u[k,ni-1]*u[k,ni-1]
    # Solve lambda
    lam[k+1,ni-1] = lam[k,ni-1] + dt*nu2/(dx*dx)*(lam[k,ni-2] - 2*lam[k,ni-1] + lam[k,0]) 
    lam[k+1,ni-1] += dt*k0*(1-lam[k,ni-1])*np.exp((u[k,ni-1]-uc)/alpha) 
    lam[k+1,ni-1] += -dt*s*up*lam[k,ni-1]/(1 + np.exp(r*(u[k,ni-1] - up)))



# plot results as contour plot
#lev = np.linspace(0,2,10)
uplot = np.matrix.transpose(u)
lplot = np.matrix.transpose(lam)

plt.contourf(uplot,levels=255,cmap='jet')
#plt.clim(0.1,100)
plt.colorbar()
#plt.contour(T,levels=24,colors='black')
plt.show()

File: 1d/ftcs_test.cpp
/*
ASTE-404 Project Test Bed
*/

#include<iostream>
#include<vector>
#include<fstream>
#include<sstream>
#include<math.h>
#include <iomanip>
#include "Matrix.h"
#include "Vec.h"


using namespace std;

struct World { // world structure
  double3 x0;	// origin
  double3 dh;   // cell spacing
  int3 nn;  	// number of nodes
  int U(int i, int j, int k=0) {return k*nn[1]*nn[0]+j*nn[0]+i;}
};


// function prototypes
vector<double> solveGS(Matrix &A, vector<double> &g, bool verbose=false);
void saveVTI(int ts, const World &world, const vector<double> &den);


int main() {

    //************** INITIALIZATION **************//

    // set world variables
    World world;
    world.x0 = {0.0,0.0,0.0};		// origin for plotting
    world.dh = {0.02,0.02,0.02};	// cell spacing
    world.nn = {161,26,1};           // number of cells
	
	int ni = world.nn[0];
	int nj = world.nn[1];
	int nu = ni*nj;

    double L = 2.16*M_PI;   // Domain length
	
	// set temporal / spatial constants
    world.dh[0] = L*1/(ni-1);
	double dx = world.dh[0];    // x spacing
	double dy = world.dh[1];    // y spacing
    double dt = 1e-4;           // time step
	int nt = 350001;			// number of timesteps

    // set physical constants
    //double L = 2*M_PI;
    double q = 1.0;
    double alpha = 0.3;
    double u_c = 1.1;
    double s = 4.5;
    double k0 = 1;
    double eps = 0.8;
    double r = 5;
    double nu1 = 0.0075;
    double nu2 = 0.0075;
    double u_p = 0.8;

    //************** END INITIALIZATION **************//



    //************** CREATE VECTORS AND MATRICES **************//

    // set solution vectors
    vector<double> u(nu);           // internal energy
    vector<double> lambda(nu);      // reactant mass fraction

    // initialize pulse characteristics
    for (int j = 0; j < nj; j++) {
        for (int i = 0; i < ni; i++) {
            int n = j*ni + i;
            u[n] = 1.5*pow((1/cosh(1*(i*dx-1))),20);
            lambda[n] = 0.5;
        }
    }

    // Coefficients to save space
    // TODO

    //************** SOLVE FTCS **************//

    // initial pressure or energy (possibly needed, arbitrary rn)
    //P = EtoP(E,rho,u,v,Y,gamma,q);
    //E = PtoE(P,rho,u,v,Y,gamma,q);

    for (int t = 0; t < nt; t++) {  // FTCS time marching

        // create copy of previous step properties
        vector<double> u_old = u;
        vector<double> lam_old = lambda;

        // Apply FTCS to internal nodes
        for (int j = 0; j < nj; j++) {
            for (int i = 0; i < ni; i++) {
                int n = j*ni + i;   // indexing
                // Apply periodic boundary conditions
                if (i == 0) {
                    // solve for energy
                    u[n] = u_old[n] + (nu1*dt/(dx*dx))*(u_old[n+1] - 2*u_old[n] + u_old[n+ni-1])
                        - (dt/(2.0*dx))*u_old[n]*(u_old[n+1] - u_old[n+ni-1])
                        + dt*k0*q*(1 - lam_old[n])*exp((u_old[n] - u_c) / alpha)
                        - dt*eps*u_old[n]*u_old[n];
                    // solve for mass fraction
                    lambda[n] = lam_old[n] + (nu2*dt/(dx*dx))*(lam_old[n+1] - 2*lam_old[n] + lam_old[n+ni-1])
                        + dt*k0*(1 - lam_old[n])*exp((u_old[n] - u_c) / alpha)
                        - dt*s*u_p*lam_old[n]/(1 + exp(r*(u_old[n] - u_p)));
                } else if (i == ni-1) {
                    // solve for energy
                    u[n] = u_old[n] + (nu1*dt/(dx*dx))*(u_old[n-(ni-1)] - 2*u_old[n] + u_old[n-1])
                        - (dt/(2.0*dx))*u_old[n]*(u_old[n-(ni-1)] - u_old[n-1])
                        + dt*k0*q*(1 - lam_old[n])*exp((u_old[n] - u_c) / alpha)
                        - dt*eps*u_old[n]*u_old[n];
                    // solve for mass fraction
                    lambda[n] = lam_old[n] + (nu2*dt/(dx*dx))*(lam_old[n-(ni-1)] - 2*lam_old[n] + lam_old[n-1])
                        + dt*k0*(1 - lam_old[n])*exp((u_old[n] - u_c) / alpha)
                        - dt*s*u_p*lam_old[n]/(1 + exp(r*(u_old[n] - u_p)));
                } else {
                    // solve for energy
                    u[n] = u_old[n] + (nu1*dt/(dx*dx))*(u_old[n+1] - 2*u_old[n] + u_old[n-1])
                        - (dt/(2.0*dx))*u_old[n]*(u_old[n+1] - u_old[n-1])
                        + dt*k0*q*(1 - lam_old[n])*exp((u_old[n] - u_c) / alpha)
                        - dt*eps*u_old[n]*u_old[n];
                    // solve for mass fraction
                    lambda[n] = lam_old[n] + (nu2*dt/(dx*dx))*(lam_old[n+1] - 2*lam_old[n] + lam_old[n-1])
                        + dt*k0*(1 - lam_old[n])*exp((u_old[n] - u_c) / alpha)
                        - dt*s*u_p*lam_old[n]/(1 + exp(r*(u_old[n] - u_p)));
                }
            }
        }

        // plot results every 100 time steps
		if (t%1000==0) { cout<<t<<endl; saveVTI(t/10,world,u);}
    }
    

	return 0;	// normal exit
}   // main end


// Gauss Seidel solver, receives references to a base Matrix type
vector<double> solveGS(Matrix &A, vector<double> &g, bool verbose) {

    int nr = A.getNr();     // number of rows

    vector<double> x(nr);   // solution vector

    /* solve matrix system */
    for (int it = 0; it < 10000; it++) { // solver iteration
        for (int r = 0; r < nr; r++) { // loop over rows
            
            double sum = A.dotRow(r,x) - A(r,r)*x[r];
            double x_star = (g[r] - sum) / A(r,r); // new estimate for x[r]

            x[r] += 1.4*(x_star-x[r]);    // SOR
        }

        // convergence check, only every 50 iterations
        if(it%50 == 0) {

            vector<double> R = A*x - g; // residue vector

            // compute average error
            double L2 = sqrt(mag2(R)/nr);

            if (verbose) {  // output iteration data if desired
                cout<<"solver iteration "<<it<<", L2 norm: "<<L2<<endl;
            }
            if (L2 < 1e-4) break; // break out of solver loop
        }
    }

    // return solution vector, hint to compiler to try moving instead of copying
    return std::move(x);
}

// writes out density field to a "results" folder (needs to be manually created!)
void saveVTI(int ts, const World &world, const vector<double> &den)  {
	// output data
	stringstream ss;
	ss<<"results/field_"<<setfill('0')<<setw(6)<<ts<<".vti";

	ofstream out(ss.str());            //open output file
	out<<"<?xml version=\"1.0\"?>\n";
	out<<"<VTKFile type=\"ImageData\">\n";
	out<<"<ImageData WholeExtent=\"0 "<<world.nn[0]-1<<" 0 "<<world.nn[1]-1<<" 0 "<<world.nn[2]-1<<"\" "
	   <<"Origin=\""<<world.x0<<"\" Spacing=\""<<world.dh<<"\">\n";
	out<<"<Piece Extent=\"0 "<<world.nn[0]-1<<" 0 "<<world.nn[1]-1<<" 0 "<<world.nn[2]-1<<"\">\n";

	out<<"<PointData>\n";
	
	out<<"<DataArray Name=\"den (#/m^3)\" NumberOfComponents=\"1\" type=\"Float64\" format=\"ascii\">\n";
	for (const double &val:den) out<<val<<" ";
	out<<"</DataArray>\n";

	out<<"</PointData>\n";

	out<<"</Piece>\n";
	out<<"</ImageData>\n";
	out<<"</VTKFile>\n";
}